openvalidators/forward.py

```diff
@@ -42,15 +42,24 @@ def get_random_uids(self, k: int, exclude: List[int] = None) -> torch.LongTensor
         If `k` is larger than the number of available `uids`, set `k` to the number of available `uids`.
     """
     candidate_uids = []
+    avail_uids = []
 
     for uid in range(self.metagraph.n.item()):
         uid_is_available = check_uid_availability(self.metagraph, uid, self.config.neuron.vpermit_tao_limit)
         uid_is_not_excluded = exclude is None or uid not in exclude
 
-        if uid_is_available and uid_is_not_excluded:
-            candidate_uids.append(uid)
+        if uid_is_available:
+            avail_uids.append(uid)
+            if uid_is_not_excluded:
+                candidate_uids.append(uid)
+
+    # Check if candidate_uids contain enough for querying, if not grab all avaliable uids
```

Comment:
So we choose k UIDs that are not:
- unavailable (absent or don't meet vpermit requirements)
- explicitly excluded

## Example
With k=50:
- If we have 50 or less candidates we sample 50 available UIDs at random
- If we have more than 50 we sample 50 from the candidates.

This means that we ignore the exclude list if there are not enough candidates. It may be better to fulfill as much of the exclude list as possible before resorting to exclude-free sampling

```python
available_uids = candidate_uids
if len(candidate_uids) < k:
    available_uids += [uid for uid in avail_uids if uid not in candidate_uids]

uids = torch.tensor(random.sample(available_uids, k), dtype=torch.int64)
 ```